### PR TITLE
Dynamic attention slicing based on available free memory

### PR DESCRIPTION
**Pull request by damian0815**
Implement a sliced AttnProcessor that dynamically picks its slice size based on available (V)RAM. Enables larger image generation and (once kulinseth/#222 is merged into pytorch) large performance gains + larger image generation on MPS

### CODE REVIEW

**Comment by keturn** [comment]
Is this equivalent to a heuristic that picks a value of 2, 4, or 8 to pass to `enable_attention_slicing`, or is it trying to do something more sophisticated than that?

**Comment by damian0815** [comment]
@keturn it's more sophisticated. the idea is to query that actual available free memory and make a decision in real time about how to slice the tensor. seems to work great on mac but we had some issues getting reliable memory reporting with cuda

**Comment by JPPhoto** [comment]
I'd love for some CUDA folks to test this latest out with large images. I'm running into an OOM error with a very large image (1536x1536) on my 12GB card but it's happening at `decode_latents()` in `diffusers_pipeline.py` and out of the scope of this PR.

File: ldm/models/diffusion/dynamic_slice_attention_processor.py [hunk 1]
```diff
@@ -84,6 +83,13 @@ def __call__(self, attn: CrossAttention, hidden_states, encoder_hidden_states=No
 
             hidden_states[start_idx:end_idx] = attn_slice
 
+            del query_slice
+            del key_slice
+            del attn_mask_slice
+            del attn_slice
+            if hidden_states.device.type == 'cuda':
+                torch.cuda.empty_cache()
```
Comment (damian0815) [inline ldm/models/diffusion/dynamic_slice_attention_processor.py:91]: this has a performance penalty - i had something like it in an earlier version and lots of folks complained so i had to take it out. does it need to be run many times per step, or is just once ok?

Comment (JPPhoto) [reply ldm/models/diffusion/dynamic_slice_attention_processor.py:91]: I tried removing that block in my testing. If I don't clear the cache out right there, it dies. I'm happy to test out an alternative if you share a diff. Maybe one idea is to check if we're over a high-water mark and then empty the cache?

Comment (JPPhoto) [reply ldm/models/diffusion/dynamic_slice_attention_processor.py:91]: This is the classic tradeoff between memory and performance...

**Comment by damian0815** [comment]
wow, ok, so there's basically zero advantage to this code over just aggressively slicing everything into as small chunks as possible (`enable_attention_slicing(slice_size='max')`). let's just do that, then.

**Comment by JPPhoto** [comment]
So do you think we should choose at runtime which to use or err on the side of safety and always pick `max`? Current implementation is `auto`.

**Comment by damian0815** [comment]
> So do you think we should choose at runtime which to use or err on the side of safety and always pick `max`? Current implementation is `auto`.

can you do a comparison with no slicing at all?

**Comment by JPPhoto** [comment]
off:
1280x1280
*OOM*
1024x1024
*OOM*
768x768
`>>   1 image(s) generated in 23.67s`
512x512
`>>   1 image(s) generated in 8.41s`

`auto`:
1280x1280
*OOM*
1024x1024
`>>   1 image(s) generated in 70.30s`
768x768
`>>   1 image(s) generated in 23.64s`
512x512
`>>   1 image(s) generated in 8.55s`

`max`:
1280x1280
`>>   1 image(s) generated in 147.21s`
1024x1024
`>>   1 image(s) generated in 65.62s`
768x768
`>>   1 image(s) generated in 23.19s`
512x512
`>>   1 image(s) generated in 9.06s`

**Comment by JPPhoto** [comment]
I didn't do time trials here, but I think it's pretty conclusive that we should be using `max` for every generation. You want to make a PR that switches that up and cite my time trials?

**Comment by damian0815** [comment]
right, so there's a definite perf hit for max and a small but not so bad one for auto. so slicing should be off whenever it can be, and only turned on when it's needed. i think your suggestion is best:
> We could use some of this logic to determine whether a particular generation should use auto or max

**Comment by JPPhoto** [comment]
Is there a perf hit for max? I'm not seeing it in that data.

**Comment by damian0815** [comment]
> Is there a perf hit for max? I'm not seeing it in that data.

max:
512x512
>>   1 image(s) generated in 9.06

off:
512x512
>>   1 image(s) generated in 8.41s

**Comment by JPPhoto** [comment]
The reason I think we need it fixed at *something* at this point is that we set it once on initialization and never change it between generations. I haven't looked to see if it's even available at the generator layer, so that would be something to consider as well.

**Comment by Millu** [comment]
@damian0815 is this applicable?

**Comment by damian0815** [comment]
don't think so

**Pull request closed without merging**